File: xrpa-orchestrator/lib/runtime/python/xrpa_runtime/reconciler/object_collection.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Callable, Generic, Type, TypeVar

from xrpa_runtime.reconciler.collection_change_types import (
    CollectionChangeEventAccessor,
    CollectionChangeType,
)
from xrpa_runtime.reconciler.data_store_interfaces import (
    FullUpdateEntry,
    IDataStoreObjectAccessor,
    IObjectCollection,
)
from xrpa_runtime.reconciler.data_store_reconciler import DataStoreReconciler
from xrpa_runtime.transport.transport_stream_accessor import TransportStreamAccessor
from xrpa_runtime.utils.memory_accessor import MemoryAccessor
from xrpa_runtime.utils.xrpa_types import ObjectAccessorInterface, ObjectUuid

ObjectAccessorType = TypeVar("ObjectAccessorType", bound=ObjectAccessorInterface)
ReconciledType = TypeVar(
    "ReconciledType", bound=IDataStoreObjectAccessor[ObjectAccessorType]
)


class ObjectCollection(IObjectCollection, Generic[ObjectAccessorType, ReconciledType]):
    def __init__(
        self,
        object_accessor_type: Type[ObjectAccessorType],
        reconciler: DataStoreReconciler,
        collection_id: int,
        inbound_field_mask: int,
        indexed_field_mask: int,
        is_local_owned: bool,
    ):
        IObjectCollection.__init__(self, reconciler, collection_id)
        self._object_accessor_type = object_accessor_type
        self._inbound_field_mask = inbound_field_mask
        self._indexed_field_mask = indexed_field_mask
        self._is_local_owned = is_local_owned
        self._objects = {}
        self._create_delegate = None

    def is_local_owned(self) -> bool:
        return self._is_local_owned

    def get_object(self, id: ObjectUuid) -> ReconciledType:
        return self._objects.get(id, None)

    def get_enumerator(self) -> list[ReconciledType]:
        return list(self._objects.values())

    def __len__(self) -> int:
        return len(self._objects)

    # these functions are for updating indexes in derived classes
    def _index_notify_create(self, obj: ReconciledType) -> None:
        pass

    def _index_notify_update(self, obj: ReconciledType, fields_changed: int) -> None:
        pass

    def _index_notify_delete(self, obj: ReconciledType) -> None:
        pass

    # these functions are for isLocalOwned=true derived classes; they typically will be exposed with public wrapper functions
    def _add_object_internal(self, obj: ReconciledType):
        if not self._is_local_owned:
            return

        id = obj.get_xrpa_id()
        self._objects[id] = obj
        obj.set_xrpa_collection(self)

        if self._indexed_field_mask != 0:
            self._index_notify_create(obj)

    def _remove_object_internal(self, id: ObjectUuid):
        if not self._is_local_owned:
            return

        obj = self._objects.get(id, None)
        if obj is None:
            return

        if self._indexed_field_mask != 0:
            self._index_notify_delete(obj)

        obj.set_xrpa_collection(None)
        del self._objects[id]

    # this function is for isLocalOwned=false derived classes; it will either be called in the constructor or exposed with a public wrapper function
    def _set_create_delegate_internal(
        self,
        delegate: Callable[
            [ObjectUuid, ObjectAccessorType, IObjectCollection], ReconciledType
        ],
    ):
        self._create_delegate = delegate

    def set_dirty(self, id: ObjectUuid, fields_changed: int):
        if (self._indexed_field_mask & fields_changed) != 0:
            obj = self._objects.get(id, None)
            if obj is not None:
                self._index_notify_update(obj, fields_changed)

    def tick(self):
        for obj in self._objects.values():
            obj.tick_xrpa()

    def write_changes(self, accessor: TransportStreamAccessor, id: ObjectUuid):
        obj = self._objects.get(id, None)
        if obj is not None:
            obj.write_ds_changes(accessor)
        elif self._is_local_owned:
            change_event = accessor.write_change_event(
                CollectionChangeEventAccessor, CollectionChangeType.DeleteObject.value
            )
            change_event.set_collection_id(self._collection_id)
            change_event.set_object_id(id)

    def prep_full_update(self, entries: list[FullUpdateEntry]):
        for obj in self._objects.values():
            timestamp = obj.prep_ds_full_update()
            if timestamp > 0:
                entries.append(
                    FullUpdateEntry(
                        obj.get_xrpa_id(),
                        self._collection_id,
                        timestamp,
                    )
                )

    def process_create(self, id: ObjectUuid, mem_accessor: MemoryAccessor):
        if self._is_local_owned:
            return

        obj_accessor = self._object_accessor_type(mem_accessor)
        obj = self._create_delegate(id, obj_accessor, self)
        if obj is None:
            return

        self._objects[id] = obj
        self._process_update_internal(id, mem_accessor, self._inbound_field_mask, False)

        if self._indexed_field_mask != 0:
            self._index_notify_create(obj)

    def process_update(
        self, id: ObjectUuid, mem_accessor: MemoryAccessor, fields_changed: int
    ) -> bool:
        return self._process_update_internal(id, mem_accessor, fields_changed, True)

    def _process_update_internal(
        self,
        id: ObjectUuid,
        mem_accessor: MemoryAccessor,
        fields_changed: int,
        notify: bool,
    ) -> bool:
        fields_changed = fields_changed & self._inbound_field_mask
        if fields_changed == 0:
            # No inbound fields changed, ignore this update
            return False

        obj = self._objects.get(id, None)
        if obj is None:
            return False

        obj_accessor = self._object_accessor_type(mem_accessor)
        obj.process_ds_update(obj_accessor, fields_changed)

        if notify and (self._indexed_field_mask & fields_changed) != 0:
            self._index_notify_update(obj, fields_changed)

        return True

    def process_delete(self, id: ObjectUuid):
        if self._is_local_owned:
            return

        obj = self._objects.get(id, None)
        if obj is not None:
            self._process_delete_internal(id, obj)

    def _process_delete_internal(self, id: ObjectUuid, obj: ReconciledType):
        if self._indexed_field_mask != 0:
            self._index_notify_delete(obj)

        obj.handle_xrpa_delete()
        del self._objects[id]

    def process_message(
        self,
        id: ObjectUuid,
        message_type: int,
        timestamp: int,
        msg_accessor: MemoryAccessor,
    ):
        obj = self._objects.get(id, None)
        if obj is None:
            return

        obj.process_ds_message(message_type, timestamp, msg_accessor)

    def process_upsert(self, id: ObjectUuid, mem_accessor: MemoryAccessor):
        if not self._process_update_internal(
            id, mem_accessor, self._inbound_field_mask, True
        ):
            self.process_create(id, mem_accessor)

    def process_full_reconcile(self, reconciled_ids: set[ObjectUuid]):
        if self._is_local_owned:
            return

        to_delete = set(self._objects.keys()) - reconciled_ids

        for id in to_delete:
            self._process_delete_internal(id, self._objects[id])

    def process_shutdown(self):
        if self._is_local_owned:
            return

        for obj in list(self._objects.values()):
            self._process_delete_internal(obj.get_xrpa_id(), obj)

File: core/runtime/cs/xrpa-runtime/Runtime/reconciler/ObjectCollection.cs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


using System.Collections.Generic;

namespace Xrpa
{

    public abstract class ObjectCollection<ObjectAccessorType, ReconciledType> : IObjectCollection, IEnumerable<ReconciledType>
      where ObjectAccessorType : ObjectAccessorInterface, new()
      where ReconciledType : class, IDataStoreObjectAccessor<ObjectAccessorType>
    {

        public delegate ReconciledType CreateDelegateFunction(
          ObjectUuid id,
          ObjectAccessorType remoteValue,
          IObjectCollection collection
        );

        private readonly ulong _inboundFieldMask;
        private readonly ulong _indexedFieldMask;
        private readonly bool _isLocalOwned;

        private readonly Dictionary<ObjectUuid, ReconciledType> _objects;
        private CreateDelegateFunction _createDelegate;

        public ObjectCollection(
          DataStoreReconciler reconciler,
          int collectionId,
          ulong inboundFieldMask,
          ulong indexedFieldMask,
          bool isLocalOwned) : base(reconciler, collectionId)
        {
            _inboundFieldMask = inboundFieldMask;
            _indexedFieldMask = indexedFieldMask;
            _isLocalOwned = isLocalOwned;
            _objects = new();
            _createDelegate = null;
        }

        public override bool IsLocalOwned()
        {
            return _isLocalOwned;
        }

        public ReconciledType GetObject(ObjectUuid id)
        {
            if (_objects.TryGetValue(id, out ReconciledType entry))
            {
                return entry;
            }
            return null;
        }

        public IEnumerator<ReconciledType> GetEnumerator()
        {
            return _objects.Values.GetEnumerator();
        }

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

        public int Count => _objects.Count;

        // these functions are for updating indexes in derived classes
        protected virtual void IndexNotifyCreate(ReconciledType obj) { }
        protected virtual void IndexNotifyUpdate(ReconciledType obj, ulong fieldsChanged) { }
        protected virtual void IndexNotifyDelete(ReconciledType obj) { }

        // these functions are for isLocalOwned=true derived classes; they typically will be exposed with
        // public wrapper functions
        protected void AddObjectInternal(ReconciledType obj)
        {
            if (!_isLocalOwned)
            {
                return;
            }

            ObjectUuid id = obj.GetXrpaId();
            _objects.Add(id, obj);
            obj.SetXrpaCollection(this);

            if (_indexedFieldMask != 0)
            {
                IndexNotifyCreate(obj);
            }
        }

        protected void RemoveObjectInternal(ObjectUuid id)
        {
            if (!_isLocalOwned)
            {
                return;
            }

            if (!_objects.ContainsKey(id))
            {
                return;
            }

            ReconciledType obj = _objects[id];
            if (_indexedFieldMask != 0)
            {
                IndexNotifyDelete(obj);
            }

            obj.SetXrpaCollection(null);
            _objects.Remove(id);
        }

        // this function is for isLocalOwned=false derived classes; it will either be called in the
        // constructor or exposed with a public wrapper function
        protected void SetCreateDelegateInternal(CreateDelegateFunction createDelegate)
        {
            _createDelegate = createDelegate;
        }

        public override void SetDirty(ObjectUuid id, ulong fieldsChanged)
        {
            if ((_indexedFieldMask & fieldsChanged) != 0)
            {
                if (_objects.TryGetValue(id, out ReconciledType obj))
                {
                    IndexNotifyUpdate(obj, fieldsChanged);
                }
            }
        }

        public override void Tick()
        {
            foreach (var obj in _objects)
            {
                obj.Value.TickXrpa();
            }
        }

        public override void WriteChanges(TransportStreamAccessor accessor, ObjectUuid id)
        {
            if (_objects.TryGetValue(id, out ReconciledType obj))
            {
                obj.WriteDSChanges(accessor);
            }
            else if (_isLocalOwned)
            {
                var changeEvent = accessor.WriteChangeEvent<CollectionChangeEventAccessor>((int)CollectionChangeType.DeleteObject);
                changeEvent.SetCollectionId(GetId());
                changeEvent.SetObjectId(id);
            }
        }

        public override void PrepFullUpdate(List<FullUpdateEntry> entries)
        {
            foreach (var obj in _objects)
            {
                ulong timestamp = obj.Value.PrepDSFullUpdate();
                if (timestamp > 0)
                {
                    entries.Add(new FullUpdateEntry(obj.Key, GetId(), timestamp));
                }
            }

        }

        public override void ProcessCreate(ObjectUuid id, MemoryAccessor memAccessor)
        {
            if (_isLocalOwned)
            {
                return;
            }

            ReconciledType obj = null;
            var objAccessor = new ObjectAccessorType();
            objAccessor.SetAccessor(memAccessor);

            // create a new object using the delegate function
            obj = _createDelegate(id, objAccessor, this);
            if (obj == null)
            {
                return;
            }

            _objects.Add(id, obj);
            ProcessUpdateInternal(id, memAccessor, _inboundFieldMask, false);

            if (_indexedFieldMask != 0)
            {
                IndexNotifyCreate(obj);
            }
        }

        public override bool ProcessUpdate(
            ObjectUuid id,
            MemoryAccessor memAccessor,
            ulong fieldsChanged)
        {
            return ProcessUpdateInternal(id, memAccessor, fieldsChanged, true);
        }

        private bool ProcessUpdateInternal(
            ObjectUuid id,
            MemoryAccessor memAccessor,
            ulong fieldsChanged,
            bool notify)
        {
            fieldsChanged &= _inboundFieldMask;
            if (fieldsChanged == 0)
            {
                // no inbound fields changed, ignore this update
                return false;
            }

            if (!_objects.TryGetValue(id, out ReconciledType obj))
            {
                return false;
            }

            var objAccessor = new ObjectAccessorType();
            objAccessor.SetAccessor(memAccessor);
            obj.ProcessDSUpdate(objAccessor, fieldsChanged);

            if (notify && (_indexedFieldMask & fieldsChanged) != 0)
            {
                IndexNotifyUpdate(obj, fieldsChanged);
            }

            return true;
        }

        public override void ProcessDelete(ObjectUuid id)
        {
            if (_isLocalOwned)
            {
                return;
            }

            if (_objects.TryGetValue(id, out ReconciledType obj))
            {
                ProcessDeleteInternal(id, obj);
            }
        }

        private void ProcessDeleteInternal(ObjectUuid id, ReconciledType obj)
        {
            if (_indexedFieldMask != 0)
            {
                IndexNotifyDelete(obj);
            }

            obj.HandleXrpaDelete();

            _objects.Remove(id);
        }

        public override void ProcessMessage(
            ObjectUuid id,
            int messageType,
            ulong timestamp,
            MemoryAccessor msgAccessor)
        {
            if (!_objects.TryGetValue(id, out ReconciledType obj))
            {
                return;
            }

            obj.ProcessDSMessage(messageType, timestamp, msgAccessor);
        }

        public override void ProcessUpsert(ObjectUuid id, MemoryAccessor memAccessor)
        {
            if (!ProcessUpdateInternal(id, memAccessor, _inboundFieldMask, true))
            {
                ProcessCreate(id, memAccessor);
            }
        }

        public override void ProcessFullReconcile(HashSet<ObjectUuid> reconciledIds)
        {
            if (_isLocalOwned)
            {
                return;
            }

            List<ObjectUuid> toDelete = new();
            foreach (var kvp in _objects)
            {
                if (!reconciledIds.Contains(kvp.Key))
                {
                    toDelete.Add(kvp.Key);
                }
            }
            foreach (var id in toDelete)
            {
                ProcessDeleteInternal(id, _objects[id]);
            }
        }

        public override void ProcessShutdown()
        {
            if (_isLocalOwned)
            {
                return;
            }
            (new List<ObjectUuid>(_objects.Keys)).ForEach(id => ProcessDeleteInternal(id, _objects[id]));
        }
    }

}
